src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMath.cs

```diff
@@ -811,20 +814,41 @@ public static FormulaValue Power(IRContext irContext, NumberValue[] args)
             return FiniteChecker(irContext, 1, result);
         }
 
-        private static FormulaValue Rand(IRContext irContext, FormulaValue[] args)
+        // Since IRandomService is a pluggable service,
+        // validate that the implementation is within spec.
+        // This catches potential host bugs. 
+        private static double SafeNextDouble(this IRandomService random)
         {
-            lock (_randomizerLock)
-            {
-                if (_random == null)
-                {
-                    _random = new Random();
-                }
+            var value = random.NextDouble(); 
 
-                return new NumberValue(irContext, _random.NextDouble());
+            if (value < 0 || value > 1)
+            {
+                // this a bug in the host's IRandomService
+                throw new InvalidOperationException($"IRandomService ({random.GetType().FullName}) returned an illegal value {value}. Must be between 0 and 1");
             }
+
+            return value;
+        }
+
+        private static double SafeNextDouble(this IServiceProvider services)
```

Comment:
>SafeNextDouble [](http://example.com/codeflow?start=30&length=14)

Now that we are allowing hosts to provide services, we need to have more validation that they're doing it correctly. Else their bugs become our bugs.